### PR TITLE
Decouple OrcPageSource from HiveColumnHandle

### PR DESCRIPTION
**Pull request by linxingyuan1102**
This is a counterpart of #1552 , also part of the efforts to remove `HiveColumnHandle` from Iceberg Connector (#1324 ). (We don't use `OrcPageSource` in Iceberg Connector now but will do in the future)

### CODE REVIEW

**Comment by linxingyuan1102** [comment]
@dain Thanks. Updated.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java [hunk 1]
```diff
@@ -134,15 +95,16 @@ public Page getNextPage()
...
-            for (int fieldId = 0; fieldId < blocks.length; fieldId++) {
-                if (constantBlocks[fieldId] != null) {
-                    blocks[fieldId] = constantBlocks[fieldId].getRegion(0, batchSize);
-                }
-                else {
-                    blocks[fieldId] = new LazyBlock(batchSize, new OrcBlockLoader(hiveColumnIndexes[fieldId]));
-                }
-            }
+            Block[] blocks = includedColumns.entrySet().stream()
+                    .map(entry -> {
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]: As a general rule, we avoid streams during query execution.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java [hunk 2]
```diff
@@ -66,43 +55,15 @@
...
-            hiveColumnIndexes[columnIndex] = column.getHiveColumnIndex();
-
-            if (!recordReader.isColumnPresent(column.getHiveColumnIndex())) {
-                constantBlocks[columnIndex] = RunLengthEncodedBlock.create(type, null, MAX_BATCH_SIZE);
-            }
-        }
-        types = typesBuilder.build();
-        columnNames = namesBuilder.build();
-
+        this.includedColumns = requireNonNull(includedColumns, "includedColumns is null");
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]:
```suggestion
        this.includedColumns = ImmutableMap.copyOf(requireNonNull(includedColumns, "includedColumns is null"));
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java [hunk 3]
```diff
@@ -214,8 +214,7 @@ public static OrcPageSource createOrcPageSource(
             return new OrcPageSource(
                     recordReader,
                     orcDataSource,
-                    physicalColumns,
-                    typeManager,
+                    includedColumns.build(),
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java]: build it once -- there already is `includedColumns.build()` above

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java [hunk 4]
```diff
@@ -66,43 +55,15 @@
...
-        this.stats = requireNonNull(stats, "stats is null");
-
-        this.constantBlocks = new Block[size];
-        this.hiveColumnIndexes = new int[size];
-
-        ImmutableList.Builder<String> namesBuilder = ImmutableList.builder();
-        ImmutableList.Builder<Type> typesBuilder = ImmutableList.builder();
-        for (int columnIndex = 0; columnIndex < columns.size(); columnIndex++) {
-            HiveColumnHandle column = columns.get(columnIndex);
-            checkState(column.getColumnType() == REGULAR, "column type must be regular");
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]: Where did this logic go? 
Am I correct this is handled by `HivePageSourceProvider`?

If so, I'd name this commit differently. Something signalling you're simply reusing information already collected (removing unnecessary code). Or, you may include this information as part of commit message body.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java [hunk 5]
```diff
@@ -134,15 +95,16 @@ public Page getNextPage()
...
-                }
-                else {
-                    blocks[fieldId] = new LazyBlock(batchSize, new OrcBlockLoader(hiveColumnIndexes[fieldId]));
-                }
-            }
+            Block[] blocks = includedColumns.entrySet().stream()
+                    .map(entry -> {
+                        int columnIndex = entry.getKey();
+                        Type type = entry.getValue();
+                        return recordReader.isColumnPresent(columnIndex) ?
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]: I'd prefer `if` here over `?:`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java [hunk 6]
```diff
@@ -134,15 +95,16 @@ public Page getNextPage()
                 return null;
             }
 
-            Block[] blocks = new Block[hiveColumnIndexes.length];
-            for (int fieldId = 0; fieldId < blocks.length; fieldId++) {
-                if (constantBlocks[fieldId] != null) {
-                    blocks[fieldId] = constantBlocks[fieldId].getRegion(0, batchSize);
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]: We're no longer reusing `constantBlocks`. I think this is OK. @dain

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java [hunk 7]
```diff
@@ -66,43 +55,15 @@
     public OrcPageSource(
             OrcRecordReader recordReader,
             OrcDataSource orcDataSource,
-            List<HiveColumnHandle> columns,
-            TypeManager typeManager,
+            Map<Integer, Type> includedColumns,
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]: I don't like the usage of `Map` here.

- it's very non-obvious what map keys are
- this is not used as a map. It's just a list of pairs `(hive column index, column type)` (... or triples `(block no, hive column index, column type)` where `block no` is implied from iteration order)

A List would be better.

I know this is pre-existing, but I think we should improve (might be a follow-up).

Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]: In https://github.com/prestosql/presto/pull/1552 you seem to be using `List<Type> types, List<Optional<Field>> fields` for similar purpose. 
I'm ok with two-list approach.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java (hunk 4 above)
Comment (linxingyuan1102) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]: The useful logic in this block is the checking done by `recordReader.isColumnPresent`. I moved it to `getNextPage`.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java (hunk 7 above)
Comment (linxingyuan1102) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]: > I don't like the usage of `Map` here.
> 
> * it's very non-obvious what map keys are
> * this is not used as a map. It's just a list of pairs `(hive column index, column type)` (... or triples `(block no, hive column index, column type)` where `block no` is implied from iteration order)
> 
> A List would be better.
> 
> I know this is pre-existing, but I think we should improve (might be a follow-up).

Your description is precise. It's essentially a list of triples. I think we can do the improvement as a follow-up.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java (hunk 4 above)
Comment (dain) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]: @findepi I had to do some digging when I reviewed this.  This code is a duplicate of https://github.com/prestosql/presto/blob/f37659ee6ce907e51a75f325ed744c2fcbc459ca/presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java#L193-L201  I think this is just cruft from refactorings.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java (hunk 7 above)
Comment (dain) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]: This is the API we already use in [`OrcRecordReader`](https://github.com/prestosql/presto/blob/f37659ee6ce907e51a75f325ed744c2fcbc459ca/presto-orc/src/main/java/io/prestosql/orc/OrcRecordReader.java#L116), so it makes sense to me.  

That said, we are going to need some better abstractions around ORC reader columns in general to deal with features like ACID, and I think we should deal with the new API with that PR instead.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java (hunk 6 above)
Comment (dain) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSource.java]: For both this and the streaming, I'm not super concerned as this code is only executed once per page.  That said, we haven't measured the impact, and the new code isn't dramatically more readable, so I'd would have left is as is to keep risk down.  I don't have a strong preference, so I open to either solution.

**Comment by linxingyuan1102** [comment]
@dain @findepi Thanks for the review. I've updated the PR.

**Pull request merged**